gtk/Session.cc

```diff
@@ -905,18 +907,62 @@ void Session::torrent_changed(tr_torrent_id_t id)
 
 void Session::remove_torrent(tr_torrent_id_t id, bool delete_files)
 {
+    struct CallbackUserData
+    {
+        std::weak_ptr<Impl> impl;
+        tr_torrent_id_t id;
+        bool succeeded = false;
+    };
+
     if (auto const& [torrent, position] = impl_->find_torrent_by_id(id); torrent)
     {
-        /* remove from the gui */
-        impl_->get_raw_model()->remove(position);
+        // This is the callback called from the Gtk main thread. If successfull,
+        // it removes the torrent entry from the GUI.
+        auto callback_this_thread = [](gpointer user_data) -> gboolean
+        {
+            auto ud = std::unique_ptr<CallbackUserData>(static_cast<CallbackUserData*>(user_data));
+            if (ud->succeeded)
+            {
+                if (auto const impl = ud->impl.lock(); impl)
+                {
+                    if (auto const& [torrent_, position_] = impl->find_torrent_by_id(ud->id); torrent_)
+                    {
+                        /* remove from the gui */
+                        impl->get_raw_model()->remove(position_);
+                    }
+                }
+            }
+
+            return G_SOURCE_REMOVE;
+        };
+        // Convert it explicitly to a constexpr C function pointer so it can be
+        // used from inside the next lambda without capturing (so that itself
+        // can be used as a C function pointer).
+        constexpr GSourceFunc CALLBACK_THIS_THREAD_PTR = callback_this_thread;
+
+        // This is the callback called from the libtransmission thread, it
+        // stores the result and schedules the callback to be called from the
+        // main thread.
+        auto callback_other_thread = [](bool succeeded, void* user_data)
+        {
+            // Store the result for the Gtk thread callback to use:
+            auto* ud = static_cast<CallbackUserData*>(user_data);
+            ud->succeeded = succeeded;
+
+            // Unfortunatelly, there is no thread-safe C++ binding to
+            // this function, so we have to use the C API directly.
+            g_idle_add(CALLBACK_THIS_THREAD_PTR, user_data);
```

Comment:
There is in fact a C++ binding, see `Glib::signal_idle()`

Reply:
`Glib::signal_idle()` is not thread-safe, and can only be called from the main context thread, which is not the case here.

Reply:
IIUC, `connect()` isn't thread-safe but `connect_once()` is (with a small caveat that doesn't apply if lambda is used as a slot), and you need the latter based on you unconditionally returning `G_SOURCE_REMOVE` from `callback_this_thread` and not tracking the signal connection.

Reply:
Actually, it is thread-safe if you don't use `sigc::mem_fun()`, but instead use a lambda: https://gnome.pages.gitlab.gnome.org/glibmm/classGlib_1_1SignalIdle.html#adeff0ddfe5eff1e0dd8302401437d8bc. So it will work!

But I couldn't get away with the explicit struct, because I have to pass both the session and the torrent id. It will be possible to remove if/when libtransmission uses a C++ interface, so the callback lambda will be allowed to capture these values.

Reply:
I did suggest (and would still like) to pass torrent id as a callback parameter, for context, so you don't need it in the struct, which leaves just the session pointer.

Reply:
Yeah, I remember. I did try it, but it didn't get much better, because I still had to do `new RefPtr<Session>{...}` and then do the casting inside the callback, so in my opinion it was not worth it. But if you insist, I can do it.